sdks/python/apache_beam/runners/worker/statesampler.pyx

```diff
@@ -180,21 +184,43 @@ cdef class StateSampler(object):
         self.scoped_states_by_index[self.current_state_index].name,
         self.state_transition_count)
 
-  def scoped_state(self, name):
-    """Returns a context manager managing transitions for a given state."""
-    cdef ScopedState scoped_state = self.scoped_states_by_name.get(name, None)
+  def scoped_state(self, step_name, state_name=None, io_target=None):
+    """Returns a context manager managing transitions for a given state.
+    Args:
+      step_name: A string with the name of the running step.
+      state_name: A string with the name of the state (e.g. 'process', 'start')
+      io_target: An IOTargetName object describing the io_target (e.g. writing
+        or reading to side inputs, shuffle or state). Will often be None.
+
+    Returns:
+      A ScopedState for the set of step-state-io_target.
+    """
+    cdef ScopedState scoped_state
+    if state_name is None:
+      # If state_name is None, the worker is still using old style
+      # msec counters.
+      counter_name = '%s-%s-msecs' % (self.prefix, step_name)
+      scoped_state = self.scoped_states_by_name.get(step_name, None)
+    else:
+      counter_name = CounterName(state_name+'-msecs',
+                                 stage_name=self.prefix,
+                                 step_name=step_name,
+                                 io_target=io_target)
+      scoped_state = self.scoped_states_by_name.get(counter_name, None)
```

Comment:
<!--new_thread; commit:b3d629fb27b7150b79f1bfcea973f5ff0c0a1b78; resolved:0-->
We are conflating the type of what can be used as keys in `self.scoped_states_by_name`.  Previously, this was just strings, but now, you are using this `CounterName` object, which doesn't seem correct.  The `CounterName` class doesn't define `__hash__()` or even `__cmp__()` / `__eq__()`, so it is possible for you to create two semantically identical CounterName objects which do not compare as equal, and are not considered the same key by the `self.scoped_states_by_name` `dict` object.

This is related to the comment above noting that sometimes `counter_name` in this code is a `CounterName`, and at other times, it is a string.  We should make this code unambiguous in this respect.

Reply:
The `CounterName` class is a subclass of a namedtuple (which, itself contains only strings or other namedtuples) - therefore it does implement the `__hash__`, `__eq__`, and `__cmp__`. So two objects that are semantically equal will hash to the same value. So this should be safe.

Does that make sense? What do you think?